### PR TITLE
refactor: code

### PR DESCRIPTION
**Pull request by evilebottnawi**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

one util for schema validation for webpack/loaders/plugins

### Breaking Changes

Yes

### Additional Info

- `not` keyword is not supported  (will be implemented in future)
- `$data` not supported (will be implemented in future)
- need improve pretty print for `const` keyword 

/cc @sokra @hiroppy

### CODE REVIEW

File: src/ValidationError.js [hunk 1]
```diff
@@ -1,23 +1,854 @@
...
+
+  return newChildren;
+}
+
+function indent(str, prefix) {
+  return str.replace(/\n(?!$)/g, `\n${prefix}`);
+}
+
+function isObject(maybyObj) {
+  return typeof maybyObj === 'object' && maybyObj !== null;
```
Comment (hiroppy) [inline src/ValidationError.js]:
```diff
- return typeof maybyObj === 'object' && maybyObj !== null;
+ return typeof maybyObj === 'object';
```

Do we need `maybeObj !== null`?

Comment (evilebottnawi) [reply src/ValidationError.js]: `typeof` null is object, so we should avoid this behavior

File: src/ValidationError.js [hunk 2]
```diff
@@ -1,23 +1,854 @@
...
+    this.schema = schema;
+    this.headerName = configuration.name || 'Object';
+    this.baseDataPath = configuration.baseDataPath || 'configuration';
+    this.postFormatter = configuration.postFormatter || null;
+
+    const header = `Invalid ${this.baseDataPath} object. ${
+      this.headerName
+    } has been initialised using ${
+      this.baseDataPath === 'options'
+        ? `an ${this.baseDataPath}`
```
Comment (sokra) [inline src/ValidationError.js]: Maybe /^[aeiou]/

File: src/ValidationError.js [hunk 3]
```diff
@@ -1,23 +1,854 @@
...
+      return `\n-> ${schemaPart.description}`;
+    }
+
+    return '';
+  }
+
+  formatValidationError(error) {
+    const dataPath = `${this.baseDataPath}${error.dataPath}`;
+
+    if (error.keyword === 'type') {
```
Comment (sokra) [inline src/ValidationError.js]: switch

File: src/ValidationError.js [hunk 4]
```diff
@@ -1,23 +1,854 @@
...
+  formatValidationErrors(errors) {
+    return errors
+      .map((error) => {
+        let formattedError = this.formatValidationError(error, this.schema);
+
+        if (this.postFormatter) {
+          formattedError = this.postFormatter(formattedError, error);
+        }
+
+        if (error.keyword === 'contains') {
```
Comment (sokra) [inline src/ValidationError.js]: Huh?

File: src/keywords/absolutePath.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+export default (ajv) =>
+  ajv.addKeyword('absolutePath', {
+    errors: true,
+    type: 'string',
+    compile(expected, schema) {
+      function callback(data) {
+        let passes = true;
+        const isExclamationMarkPresent = data.includes('!');
+        const isCorrectAbsoluteOrRelativePath =
+          expected === /^(?:[A-Za-z]:\\|\/)/.test(data);
```
Comment (sokra) [inline src/keywords/absolutePath.js]: A recent webpack PR adds windows network shares as absolute path starting with `\\`

Comment (evilebottnawi) [reply src/keywords/absolutePath.js]: https://github.com/webpack/webpack/pull/9382? Looks not merged right now

File: src/ValidationError.js [hunk 6]
```diff
@@ -1,23 +1,854 @@
...
+      )}.${this.getSchemaPartDescription(error.parentSchema)}`;
+    } else if (
+      error.keyword === 'formatMinimum' ||
+      error.keyword === 'formatMaximum'
+    ) {
+      return `${dataPath} should be ${error.params.comparison} ${JSON.stringify(
+        error.params.limit
+      )}.${this.getSchemaPartDescription(error.parentSchema)}`;
+    } else if (
+      error.keyword === 'minimum' ||
```
[the remark this answers is not in the document]
Comment (evilebottnawi) [reply src/ValidationError.js]: I think no, `formatMinimum` and `formatMaximum` require `JSON.stringify(error.params.limit)`, but `minimum`/`maximum`/`exclusiveMinimum`/`exclusiveMaximum` support only numbers therefore no need to use `JSON.stringify`

File: src/ValidationError.js [hunk 7]
```diff
@@ -1,23 +1,854 @@
...
+    if (schema.anyOf) {
+      return schema.anyOf.map(formatInnerSchema).join(' | ');
+    }
+
+    if (schema.allOf) {
+      return schema.allOf.map(formatInnerSchema).join(' & ');
+    }
+
+    if (schema.if) {
+      return `if the data is valid against the sub-schema ${formatInnerSchema(
```
[the remark this answers is not in the document]
Comment (evilebottnawi) [reply src/ValidationError.js]: yep, we simplify this, but for objects it is looks some misleading:
```
if object { power, … } then object { disbelief, … } else object { confidence, … }
```
Maybe we should output better message

File: src/keywords/absolutePath.js (hunk 5 above)
Comment (sokra) [reply src/keywords/absolutePath.js]: tests are missing, but `\\computer\share` is an absolute path for sure.

It's difficult to test inside of webpack, but should be easy to test here.

Comment (evilebottnawi) [reply src/keywords/absolutePath.js]: :+1: so i just move this here and test

File: src/ValidationError.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sokra) [reply src/ValidationError.js]: oh, I didn't spot this difference. Fine then.

**Comment by sokra** [comment]
Awesome, go with it...

**Pull request merged**